overlord/snapstate/storehelpers.go

```diff
@@ -422,6 +455,36 @@ func updateToRevisionInfo(st *state.State, snapst *SnapState, revision snap.Revi
 		Revision: revision,
 	}
 
+	enforcedSets, err := EnforcedValidationSets(st)
+	if err != nil {
+		return nil, err
+	}
+	if enforcedSets != nil {
+		requiredValsets, requiredRevision, err := enforcedSets.CheckPresenceRequired(naming.Snap(curInfo.InstanceName()))
+		if err != nil {
+			return nil, err
+		}
+		if !requiredRevision.Unset() {
+			if revision != requiredRevision {
+				return nil, fmt.Errorf("cannot update snap %q to revision %s without --ignore-validation, revision %s is required by validation sets: %s",
+					curInfo.InstanceName(), revision, requiredRevision, strings.Join(requiredValsets, ","))
```

Comment:
Is it correct to do this even if `requiredValsets` is empty, or should this `if` block be merged together with the next one?

Reply:
It is correct, requiredRevision will only be set if there are requiredValsets (in other words, required revision number comes from validation sets that require it). Of course requiredRevision may be unset even if we have requiredValsets here - it depends on whether the validation set assertions have the optional "required: <revision>" attribute for the given snap or not.